src/Linters/PackagePrivateLinter.hack

```diff
@@ -0,0 +1,318 @@
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace Facebook\HHAST;
+
+use type Facebook\HHAST\{
+  ASTLintError,
+  ASTLinter,
+  FunctionToken,
+  INamespaceUseDeclaration,
+  NameToken,
+  NamespaceDeclaration,
+  NamespaceGroupUseDeclaration,
+  NamespaceToken,
+  NamespaceUseClause,
+  NamespaceUseDeclaration,
+  Node,
+  QualifiedName,
+  Script,
+  TypeToken,
+};
+use namespace HH\Lib\{C, Str, Vec};
+
+type use_definition_parts_t = shape(
+  'parts' => vec<string>,
+  'alias' => ?string,
+  'node' => Node,
+);
+type package_private_use_definitions_t = shape(
+  'function' => vec<use_definition_parts_t>,
+  'namespace' => dict<string, use_definition_parts_t>,
+  'type' => vec<use_definition_parts_t>,
+);
+
+final class PackagePrivateLinter extends ASTLinter {
+  const type TContext = Script;
+  const type TNode = Script;
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    this::TContext $context,
+    this::TNode $node,
+  ): ?ASTLintError {
+    // Get the current namespace for the file.
+    $current_namespace_for_file = vec[];
+    $top_level_namespace = $this->getTopLevelNamespace($node);
+    if ($top_level_namespace is nonnull) {
+      $current_namespace_for_file = Str\split($top_level_namespace, '\\');
+    }
+
+    // Process all use statements present in the file.
+    $use_definition_context_shape = $this->getUseDefinitions($context);
```

Comment:
If you need something low level, `Facebook\HHAST\__Private\get_uses_directly_in_scope` - but I think `$script->getNamespaces()` gives you everything you need here without needing that; it can also reimplement getToplevelNamespace

Reply:
This was a game changer, thank you!